internal/awsdata/kms.go

```diff
@@ -0,0 +1,88 @@
+package awsdata
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/kms"
+	"github.com/manywho/awsinventory/internal/inventory"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// AssetTypeKMSKey is the value used in the AssetType field when fetching KMS keys
+	AssetTypeKMSKey string = "KMS Key"
+
+	// ServiceKMS is the key for the KMS service
+	ServiceKMS string = "kms"
+)
+
+func (d *AWSData) loadKMSKeys(region string) {
+	defer d.wg.Done()
+
+	kmsSvc := d.clients.GetKMSClient(region)
+
+	log := d.log.WithFields(logrus.Fields{
+		"region":  region,
+		"service": ServiceKMS,
+	})
+
+	log.Info("loading data")
+
+	var keys []*kms.KeyListEntry
+	done := false
+	params := &kms.ListKeysInput{}
+	for !done {
+		out, err := kmsSvc.ListKeys(params)
+
+		if err != nil {
+			d.results <- result{Err: err}
+			return
+		}
+
+		keys = append(keys, out.Keys...)
+
+		if aws.BoolValue(out.Truncated) {
+			params.Marker = out.NextMarker
+		} else {
+			done = true
+		}
+	}
+
+	log.Info("processing data")
+
+	for _, k := range keys {
+		out, err := kmsSvc.DescribeKey(&kms.DescribeKeyInput{
+			KeyId: k.KeyId,
+		})
+		if err != nil {
+			d.results <- result{Err: err}
+			return
+		}
+
+		var comments []string
+		comments = append(comments, fmt.Sprintf("%s, %s", aws.StringValue(out.KeyMetadata.KeyManager), aws.StringValue(out.KeyMetadata.CustomerMasterKeySpec)))
+		comments = append(comments, "Created at: "+aws.TimeValue(out.KeyMetadata.CreationDate).Format(time.RFC3339))
+		if !out.KeyMetadata.ValidTo.IsZero() {
```

Comment:
This is method is fine apart from `ValidTo` can be nil and I get a segfault. I think you just need to add `out.KeyMetadata.ValidTo != nil` to the condition before the `IsZero()` part

Reply:
Huh. I had it that way originally, and I purposely changed it to use `IsZero()`, as when I was testing it with `nil`, I was getting `cannot use nil as type time.Time`.

https://golangcode.com/checking-if-date-has-been-set/
https://stackoverflow.com/questions/23051973/what-is-the-zero-value-for-time-time-in-go

https://golang.org/pkg/time/ states:
> The zero value of type Time is January 1, year 1, 00:00:00.000000000 UTC. As this time is unlikely to come up in practice, the IsZero method gives a simple way of detecting a time that has not been initialized explicitly.

I'll re-add the nil check, but 🤷 .

Reply:
I think it's because it's a pointer to a `time.Time` and pointers can be `nil` so you have to first check if the pointer is `nil` and then if it's the zero value afterwards. I guess in theory it _should_ either be `nil` or a non-zero `time.Time` but it's best to keep that `IsZero` check in there just in case.